torchvision/datasets/_optical_flow.py

```diff
@@ -0,0 +1,149 @@
+import os
+from abc import ABC, abstractmethod
+from glob import glob
+from pathlib import Path
+
+import numpy as np
+import torch
+from PIL import Image
+
+from ..io.image import _read_png_16
+from .vision import VisionDataset
+
+
+__all__ = (
+    "KittiFlow",
+    "Sintel",
+)
+
+
+class FlowDataset(ABC, VisionDataset):
+    def __init__(self, root, transforms=None):
+
+        super().__init__(root=root)
+        self.transforms = transforms
+
+        self._flow_list = []
+        self._image_list = []
+
+    def _read_img(self, file_name):
+        return Image.open(file_name)
+
+    @abstractmethod
+    def _read_flow(self, file_name):
+        # Return the flow or a tuple (flow, valid) for datasets where the valid mask is built-in
+        pass
+
+    def __getitem__(self, index):
+        # Some datasets like Kitti have a built-in valid mask, indicating which flow values are valid
+        # For those we return (img1, img2, flow, valid), and for the rest we return (img1, img2, flow),
+        # and it's up to whatever consumes the dataset to decide what `valid` should be.
+
+        img1 = self._read_img(self._image_list[index][0])
+        img2 = self._read_img(self._image_list[index][1])
+        flow = self._read_flow(self._flow_list[index]) if self._flow_list else None
+
+        if isinstance(flow, tuple):
+            flow, valid = flow
+        else:
+            valid = None
+
+        if self.transforms is not None:
+            img1, img2, flow, valid = self.transforms(img1, img2, flow, valid)
+
+        if valid is None:
+            return img1, img2, flow
+        else:
+            return img1, img2, flow, valid
```

Comment:
leaving the `valid` mask aside (see other discussion above for that) and focusing on the `flow` value:

- For Sintel and other datasets (FlyingThings, FlyingChairs), the flow is currently a numpy array with shape (H, W, 2).
- For Kitti, the flow is a **tensor** with shape (2, H, W)

Should I change the output of Kitti to (H, W, 2) for it to be similar to what the other datasets return? Should I change the shape *and* the type (i.e. convert the tensor to a numpy array), or just the shape?

What the model consumes eventually is a tensor with shape (2, H, W), i.e. the format that Kitti outputs natively. I feel like it might be worth it for consistency sake tho.

Reply:
Good question. Intuitively I'm weighting towards providing a consistent result (doesn't matter the option), but I think I would love to hear from @pmeier on whether this dilemma popped up in the past and how it was handled.

Reply:
> Intuitively I'm weighting towards providing a consistent result (doesn't matter the option)

Yes, that is the way to go. If there is a format that is used by the later models, go for that and convert everything to it. This is similar to images in conjunction with `transforms.ToTensor()`: the output will always be `CxHxW` regardless of how it is stored in the dataset.

Reply:
I agree. I would return consistent shapes and types, with a preference for `CxHxW` as this will work out of the box with our current transforms. We can always permute it back when using it with `grid_sample`